src/core/brownian_inline.hpp

```diff
@@ -0,0 +1,146 @@
+/*
+  Copyright (C) 2010-2018 The ESPResSo project
+  Copyright (C) 2002,2003,2004,2005,2006,2007,2008,2009,2010
+    Max-Planck-Institute for Polymer Research, Theory Group
+
+  This file is part of ESPResSo.
+
+  ESPResSo is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  ESPResSo is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** \file brownian_inline.hpp */
+
+#ifndef BROWNIAN_INLINE_HPP
+#define BROWNIAN_INLINE_HPP
+
+#include "thermostat.hpp"
+
+#ifdef BROWNIAN_DYNAMICS
+/** Propagate position: viscous drag driven by conservative forces.*/
+/*********************************************************/
+/** \name bd_drag */
+/*********************************************************/
+/**(Eq. (14.39) T. Schlick, https://doi.org/10.1007/978-1-4419-6351-2 (2010))
+ * @param &p              Reference to the particle (Input)
+ * @param dt              Time interval (Input)
+ */
+inline void bd_drag(Particle &p, double dt) {
+  // The friction tensor Z from the Eq. (14.31) of Schlick2010:
+  Thermostat::GammaType local_gamma;
+
+  if (p.p.gamma >= Thermostat::GammaType{}) {
+    local_gamma = p.p.gamma;
+  } else {
+    local_gamma = langevin_gamma;
+  }
+
+  for (int j = 0; j < 3; j++) {
+#ifdef EXTERNAL_FORCES
+    if (!(p.p.ext_flag & COORD_FIXED(j)))
```

Comment:
I realised that you also have to check whether the particle is virtual.

Reply:
thanks, I agree. Working on that.

Reply:
@hmenke thanks, I've probably done this now. I've modified the file integrate.cpp, keeping the same approach as it is done for the Langevin dynamics. Corresponding changes however are implemented in the end functions for the Brownian thermostat acting on the virtual sites (same logic as for Langevin dynamics is kept). See commits "Virtual sites propagation only for rotations" and "Virtual sites translational thermostat control".